FIAT/hdiv_trace.py

```diff
@@ -119,87 +173,168 @@ def tabulate(self, order, points, entity=None):
-            return phivals
+        # If the user is directly specifying cell-wise tabulation, return
+        # TraceErrors in dict for appropriate handling in the form compiler
+        if entity_dim not in self.dg_elements:
+            for key in phivals:
+                msg = "The HDivTrace element can only be tabulated on facets."
+                phivals[key] = TraceError(msg)
 
         else:
             # Retrieve function evaluations (order = 0 case)
-            nonzerovals = list(self.facet_element.tabulate(0, points).values())[0]
-            phivals[evalkey][nf*entity_id:nf*(entity_id + 1), :] = nonzerovals
-
-            # If asking for gradient evaluations, insert TraceError in gradient evaluations
+            # NOTE: Depending on the cell, we need to off set
+            # the facet index when tabulating particular facet
+            # entities
+            offset = 0
+            for facet_dim in sorted(self.dg_elements):
+                element = self.dg_elements[facet_dim]
+                nf = element.space_dimension()
+                num_facets = len(self.ref_el.get_topology()[facet_dim])
+
+                # Loop over the number of facets until we find a facet
+                # with matching dimension and id
+                for i in range(num_facets):
+                    # Found it!
+                    if facet_dim == entity_dim and i == entity_id:
+                        nonzerovals, = itervalues(element.tabulate(0, points))
+                        phivals[evalkey][offset:offset+nf, :] = nonzerovals
+
+                    offset += nf
+
+            # If asking for gradient evaluations, insert TraceError in
+            # gradient slots
             if order > 0:
-                for key in phivals.keys():
+                msg = "Gradients on trace elements are not well-defined."
+                for key in phivals:
                     if key != evalkey:
-                        phivals[key] = TraceError("Gradient evaluations are illegal on trace elements.")
-            return phivals
+                        phivals[key] = TraceError(msg)
+
+        return phivals
 
     def value_shape(self):
         """Return the value shape of the finite element functions."""
-        return self.facet_element.value_shape()
+        return ()
 
     def dmats(self):
         """Return dmats: expansion coefficients for basis function
         derivatives."""
-        raise NotImplementedError("dmats not implemented for the trace element.")
+        raise NotImplementedError("Not implemented for the trace element.")
 
     def get_num_members(self, arg):
         """Return number of members of the expansion set."""
-        raise NotImplementedError("get_num_members not implemented for the trace element.")
+        raise NotImplementedError("Not implemented for the trace element.")
+
+
+def construct_dg_element(ref_el, spanning_degrees):
```

Comment:
BTW why is it called a spanning degree? That term reminds me of `tsfc.ufl_utils.spanning_degree` but that spanning degree does not mean what this degree is.